homeassistant/components/dhcp/__init__.py

```diff
@@ -0,0 +1,159 @@
+"""The dhcp integration."""
+
+import fnmatch
+import logging
+from threading import Event, Thread
+
+from scapy.error import Scapy_Exception
+from scapy.layers.dhcp import DHCP
+from scapy.layers.l2 import Ether
+from scapy.sendrecv import sniff
+
+from homeassistant.const import EVENT_HOMEASSISTANT_STARTED, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import format_mac
+from homeassistant.loader import async_get_dhcp
+
+from .const import DOMAIN
+
+FILTER = "udp and (port 67 or 68)"
+REQUESTED_ADDR = "requested_addr"
+MESSAGE_TYPE = "message-type"
+HOSTNAME = "hostname"
+MAC_ADDRESS = "macaddress"
+IP_ADDRESS = "ip"
+DHCP_REQUEST = 3
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up the dhcp component."""
+
+    async def _initialize(_):
+        dhcp_watcher = DHCPWatcher(hass, await async_get_dhcp(hass))
+        dhcp_watcher.start()
+
+        def _stop(*_):
+            dhcp_watcher.stop()
+            dhcp_watcher.join()
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _stop)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, _initialize)
+    return True
+
+
+class DHCPWatcher(Thread):
+    """Class to watch dhcp requests."""
+
+    def __init__(self, hass, integration_matchers):
+        """Initialize class."""
+        super().__init__()
+
+        self.hass = hass
+        self.name = "dhcp-discovery"
+        self._integration_matchers = integration_matchers
+        self._address_data = {}
+        self._stop_event = Event()
+
+    def stop(self):
+        """Stop the thread."""
+        self._stop_event.set()
+
+    def run(self):
+        """Start watching for dhcp packets."""
+        try:
+            sniff(
+                filter=FILTER,
+                prn=self.handle_dhcp_packet,
+                stop_filter=lambda _: self._stop_event.is_set(),
+            )
+        except (Scapy_Exception, OSError) as ex:
```

Comment:
This error will also happen when Home Assistant is not run as root. This is the case when Home Assistant runs in a virtual environment. 

If we do run as root, we should probably make the logger an error. In a venv it's an info.

Reply:
If euid is 0 error else info

Reply:
addressed here https://github.com/home-assistant/core/pull/45157